### PR TITLE
Course Exercise Overview in Course Adminstration

### PR DESCRIPTION
**Pull request by sebina**
### Checklist
- [x] I've run `yarn run webpack:build:main` from the root directory to see that the project builds without errors.
- [x] I've removed unnecessary whitespace changes.
- [x] I've updated the documentation and models if necessary.
- [x] I've tested the changes and all related features on the Artemis test server

### Motivation and Context
Currently instructors have to click on a drop down button to get to the quiz overview for one course. If there is only one course the button forces the instructor to scroll down whiten the button to get to the specific type of the exercise. 
To avoid that and make it easier to understand we decided to implement an overview of the exercise for the whole course. By clicking on the exercise button the instructor is immediately navigated to all exercise type and all their corresponding exercises.

### Description
We changed the drop-down button into a normal button which navigated to the course overview page. On this page all the exercise are being displayed grouped by their exercise type. 
The instructors has the possibility to blend out some of the exercise types. 

#### Screenshots

screenshot 2019-02-23 at 15 18 49
screenshot 2019-02-23 at 15 19 01

### CODE REVIEW

**Comment by krusche** [comment]
Can we move the exercise type in the header of the card, e.g. 'Quiz Exercises', to the left and the element to open and close the card (>) to the right?

When opening the page in the beginning, it would be good, if all exercise types are already open

File: src/main/webapp/app/entities/text-exercise/text-exercise.module.ts [hunk 1]
```diff
@@ -16,6 +16,7 @@ import {
     TextExerciseService
 } from './';
 import { SortByModule } from '../../components/pipes';
+import { ProgrammingExerciseComponent } from 'app/entities/programming-exercise';
```
Comment (krusche) [inline src/main/webapp/app/entities/text-exercise/text-exercise.module.ts]: this import seems to be wrong

File: src/main/webapp/app/entities/quiz-exercise/quiz-exercise.component.html [hunk 2]
```diff
@@ -2,13 +2,13 @@
     <h4>
         <span *ngIf="course">{{course.title}} - </span><span *ngIf="quizExercises">{{quizExercises.length}} </span>
         <span jhiTranslate="arTeMiSApp.quizExercise.home.title">Quiz Exercises</span>
-        <button *ngIf="courseId" class="btn mr-2 btn-primary float-right jh-create-entity create-quiz-exercise" [routerLink]="['new']">
+        <button *ngIf="courseId" class="btn mr-2 btn-primary float-right jh-create-entity create-quiz-exercise" [routerLink]="['quiz-exercise' , 'new']">
             <span class="fa fa-plus"></span>
```
Comment (krusche) [inline src/main/webapp/app/entities/quiz-exercise/quiz-exercise.component.html]: why did you change this?

File: src/main/webapp/app/entities/programming-exercise/programming-exercise.component.html [hunk 3]
```diff
@@ -8,7 +8,7 @@ <h4>
             Link new Programming Exercise
             </span>
         </button>
-        <button *ngIf="course" id="jh-create-entity" class="btn btn-primary float-right jh-create-entity create-programming-exercise mr-1" [routerLink]="['new']">
```
Comment (krusche) [inline src/main/webapp/app/entities/programming-exercise/programming-exercise.component.html]: why did you change this?

File: src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise.module.ts [hunk 4]
```diff
@@ -16,6 +16,7 @@ import {
     FileUploadExerciseUpdateComponent
 } from './';
 import { SortByModule } from '../../components/pipes';
+import { ProgrammingExerciseComponent } from 'app/entities/programming-exercise';
```
Comment (krusche) [inline src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise.module.ts]: import seems to be wrong

**Comment by krusche** [comment]
There is a lot of duplicated code in course-overview.component.html
Please avoid code duplication and rather use a for loop through all exercise types (see e.g. the enum in the exercise.model.ts)

I also suggest to rename course-overview.component.html to course-exercises-overview.component.html

File: src/main/webapp/app/entities/text-exercise/text-exercise.module.ts (hunk 1 above)
Comment (sebina) [reply src/main/webapp/app/entities/text-exercise/text-exercise.module.ts]: Yes - I deleted the import

File: src/main/webapp/app/entities/quiz-exercise/quiz-exercise.component.html (hunk 2 above)
Comment (sebina) [reply src/main/webapp/app/entities/quiz-exercise/quiz-exercise.component.html]: This change is needed also for the quiz in other case the routing does not work with the create button

File: src/main/webapp/app/entities/programming-exercise/programming-exercise.component.html (hunk 3 above)
Comment (sebina) [reply src/main/webapp/app/entities/programming-exercise/programming-exercise.component.html]: This change is needed also for the quiz in other case the routing does not work with the create button

File: src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise.module.ts (hunk 4 above)
Comment (sebina) [reply src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise.module.ts]: Yes - I deleted it

**Comment by jpbernius** [comment]
Going back to my comment above, I think we can simplify the Exercise Overview Code  a lot by doing this:

Move this code Snippet to a new `ExerciseOverviewCardComponent`: 
```html
    <div class="question card">
         <div class="question-options card-header question-card-header">
             <h5><span jhiTranslate="arTeMiSApp.course.programmingExercises">Programming Exercises</span> </h5>
             <button
                 class="btn btn-light question-collapse"
                 (click)="isProgrammingQuestionCollapsed = !isProgrammingQuestionCollapsed"
                 [attr.aria-expanded]="!isProgrammingQuestionCollapsed"
                 [attr.aria-controls]="'collapseQuestion'">
                 <span class="fa fa-2x" [ngClass]="isProgrammingQuestionCollapsed ? 'fa-angle-right' : 'fa-angle-down'" ></span>
             </button>
         </div>
         <div class="card-body question-card-body"  [ngbCollapse]="isProgrammingQuestionCollapsed">
             <div>
                 <jhi-programming-exercise>
                 </jhi-programming-exercise>
             </div>
         </div>
     </div>
```

But do the following:
```html
    <div class="question card">
         <div class="question-options card-header question-card-header">
             <ng-content select="h5"></ng-content>
             <button
                 class="btn btn-light question-collapse"
                 (click)="isProgrammingQuestionCollapsed = !isProgrammingQuestionCollapsed"
                 [attr.aria-expanded]="!isProgrammingQuestionCollapsed"
                 [attr.aria-controls]="'collapseQuestion'">
                 <span class="fa fa-2x" [ngClass]="isProgrammingQuestionCollapsed ? 'fa-angle-right' : 'fa-angle-down'" ></span>
             </button>
         </div>
         <div class="card-body question-card-body"  [ngbCollapse]="isProgrammingQuestionCollapsed">
             <ng-content></ng-content>
         </div>
     </div>
```

The `ng-content` tag allows to project content into a Component.
See this post here for more details: [**ng-content: The hidden docs**](https://medium.com/claritydesignsystem/ng-content-the-hidden-docs-96a29d70d11b)


You can then use this new `<jhi-exercise-overview-card>` component like this:
```html
    <jhi-exercise-overview-card>
         <h5><span jhiTranslate="arTeMiSApp.course.programmingExercises">Programming Exercises</span> </h5>
         <jhi-programming-exercise></jhi-programming-exercise>
    </jhi-exercise-overview-card>
```

All the Collapse Variables can go away, as there is just one in the new component.
The Template becomes more clear as all the collapse logic is moved to the new component.

**Comment by sebina** [comment]
@krusche @jpbernius please check now - can we merge ?

File: src/main/webapp/app/entities/course/course.model.ts [hunk 5]
```diff
@@ -13,6 +21,7 @@ export class Course implements BaseEntity {
     public endDate: Moment;
     public onlineCourse = false; // default value
     public exercises: Exercise[];
+    public exerciseType: ExerciseType;
```
Comment (jpbernius) [inline src/main/webapp/app/entities/course/course.model.ts]: What is this property? Can this (and the enum above) be deleted?

Comment (krusche) [reply src/main/webapp/app/entities/course/course.model.ts]: I agree, this makes no sense here and is already defined in the exercise itself. Please delete!

File: src/main/webapp/app/entities/course/course-exercises-overview.component.html [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    <jhi-modeling-exercise [showHeading]="false"></jhi-modeling-exercise>
+</jhi-course-exercise-card>
+
+<jhi-course-exercise-card headingJhiTranslate="arTeMiSApp.course.fileUploadExercises">
+    <jhi-file-upload-exercise [showHeading]="false"></jhi-file-upload-exercise>
+</jhi-course-exercise-card>
+
+<jhi-course-exercise-card headingJhiTranslate="arTeMiSApp.course.textExercises">
+    <jhi-text-exercise [showHeading]="false"></jhi-text-exercise>
+</jhi-course-exercise-card>
```
Comment (krusche) [inline src/main/webapp/app/entities/course/course-exercises-overview.component.html]: Why is [showHeading] always false?

File: src/main/webapp/app/entities/course/course.route.ts [hunk 7]
```diff
@@ -50,6 +51,18 @@ export const courseRoute: Routes = [
...
         canActivate: [UserRouteAccessService]
     },
+    {
+        path: 'course/:courseId',
+        component: CourseExercisesOverviewComponent,
+        resolve: {
+            course: CourseResolve
+        },
+        data: {
+            authorities: ['ROLE_INSTRUCTOR', 'ROLE_ADMIN'],
```
Comment (krusche) [inline src/main/webapp/app/entities/course/course.route.ts]: Is this the route to see the new exercise overview? If yes, tutors should also see it

File: src/main/webapp/app/entities/modeling-exercise/modeling-exercise.module.ts [hunk 8]
```diff
@@ -16,38 +16,29 @@ import {
 } from './';
 import { SortByModule } from '../../components/pipes';
 import { FormDateTimePickerModule } from '../../shared/dateTimePicker/date-time-picker.module';
+import { ProgrammingExerciseComponent } from 'app/entities/programming-exercise';
```
Comment (krusche) [inline src/main/webapp/app/entities/modeling-exercise/modeling-exercise.module.ts]: this import is wrong. Please check all imports again

**Comment by krusche** [comment]
The course title on the new exercise page is missing. It would also be interesting to see how many exercises the course has in total and how many exercises each exercise category has

File: src/main/webapp/app/entities/course/course.component.html [hunk 9]
```diff
@@ -41,35 +41,9 @@ <h4>
...
-                                        jhiTranslate="arTeMiSApp.course.textExercises"
-                                ></button>
-                                <button class="dropdown-item"
-                                        routerLink="{{course.id}}/file-upload-exercise"
-                                        jhiTranslate="arTeMiSApp.course.fileUploadExercises"
-                                ></button>-->
-                            </div>
-                        </div>
+                        <button type="submit" [routerLink]="['/course', course.id ]" class="btn btn-dark btn-sm mr-1">
+                            <span class="d-none d-md-inline">Exercises</span>
```
Comment (krusche) [inline src/main/webapp/app/entities/course/course.component.html]: Please include translation and use the same design as the old Exercise button, i.e. with the same icon and the sam color:
image

**Comment by krusche** [comment]
Can you please merge all changes from develop branch, so that we can test on Bamboo that this feature also works with the latest official version. Thank you!

**Comment by krusche** [comment]
When I open the exercise overview, the browser executes 10 network requests now, some of them which are redundant
image

Can we pass the course object from the previous page to the exercise overview (or at least only load the course once and pass it to all sub pages then? (and only load it in the sub page if it is not available to stay compatible) to reduce the number of calls.

I think we should also remove the paging mechanism from the exercise calls, because it is not used anyway (as far as I know)

**Comment by jpbernius** [comment]
@sebina Please check all buttons:
On this page: https://artemistest.ase.in.tum.de/#/course/1
The Edit Button does not work for Quizes:
```
Error: Uncaught (in promise): Error: Cannot match any routes. URL Segment: 'course/1/edit/14' 
```

**Comment by sebina** [comment]
@jpbernius can you please tell me if it is okey to have one loadForCourse() method where I request all the exercises I have for this course ?

File: src/main/webapp/app/entities/course/course-exercises-overview.component.ts [hunk 10]
```diff
@@ -1,7 +1,112 @@
...
+    }
+
+    load() {
+        this.subscription = this.route.params.subscribe(params => {
+            this.courseId = params['courseId'];
+            this.loadForCourse(this.courseId);
+        });
+    }
+
+    loadForCourse(courseId: number) {
```
Comment (jpbernius) [inline src/main/webapp/app/entities/course/course-exercises-overview.component.ts]: This is nice.
But where are we using this data?
We are storing it in the `CourseExercisesOverviewComponent` but are not passing it to the exercise components as Inputs in the `course-exercises-overview.component.html` file.

What we are missing is something like this:
```html
<jhi-course-exercise-card headingJhiTranslate="arTeMiSApp.course.programmingExercises">
     <jhi-programming-exercise [showHeading]="false" [course]="course" [exercises]="programmingExercises"></jhi-programming-exercise>
 </jhi-course-exercise-card>
```

At this stage, it would be nice to introduce an abstract component class, which all exercise components inherit from.
This component can declare the `@Input()`s and also define the logic required by these components to only fetch the exercises themselves if they did not get it via the Input.

**Comment by sebina** [comment]
Made the changes and changed the back button - but now it takes way to long to load the page - should the method be changed again?  @jpbernius

**Comment by krusche** [comment]
When I navigate after opening this page, I receive an error in the console
image

**Comment by krusche** [comment]
For some reason, quiz exercises are not displayed at all. 
image

**Comment by krusche** [comment]
**Please do test changes in features branches more thoroughly before opening a pull request or requesting feedback**

It is a pity if you start to test and the most basic functionality is not working :-(

**Comment by sebina** [comment]
@krusche @jpbernius please review

File: src/main/webapp/app/entities/course/course-exercise-card.component.html [hunk 11]
```diff
@@ -0,0 +1,12 @@
+<div class="question card">
+    <div class="question-options card-header question-card-header">
+        <h5 [routerLink]="exerciseRouterLink"><span [jhiTranslate]="headingJhiTranslate"></span></h5>
+        <button class="btn btn-light question-collapse" (click)="isCollapsed = !isCollapsed"
+                [attr.aria-expanded]="!isCollapsed" [attr.aria-controls]="'collapseQuestion'">
+            <span class="fa fa-2x" [ngClass]="isCollapsed ? 'fa-angle-right' : 'fa-angle-down'"></span>
```
Comment (rpadovani) [inline src/main/webapp/app/entities/course/course-exercise-card.component.html]: Please, merge `develop` in your branch and use the `<fa-icon>` component :)

You can see more details here, with some examples of replacements: https://github.com/ls1intum/ArTEMiS/pull/157

File: src/main/webapp/app/entities/course/course-exercises-overview.component.html [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+    </jhi-course-exercise-card>
+
+    <jhi-course-exercise-card headingJhiTranslate="arTeMiSApp.course.textExercises">
+        <jhi-text-exercise [showHeading]="true" [course]="course" [textExercises]="textExercises" ></jhi-text-exercise>
+    </jhi-course-exercise-card>
+</ng-container>
+
+<ng-template #loading>
+    <h2>
+        <span style="color:grey"><span class="fa fa-spinner fa-spin"></span>&nbsp;<span>Please wait while ArTEMiS is finding the course and exercises!</span></span>
```
Comment (rpadovani) [inline src/main/webapp/app/entities/course/course-exercises-overview.component.html]: As before, please use `<fa-icon>`

File: src/main/webapp/app/entities/quiz-exercise/quiz-exercise.component.html [hunk 13]
```diff
@@ -1,20 +1,34 @@
...
+        </ng-container>
+        <button *ngIf="!routerContainPath()" class="btn mr-2 btn-primary float-right jh-create-entity create-quiz-exercise" [routerLink]="['quiz-exercise' , 'new']">
             <span class="fa fa-plus"></span>
             <span jhiTranslate="arTeMiSApp.quizExercise.home.createLabel">
                 Create new Quiz
             </span>
         </button>
-        <button *ngIf="courseId" class="btn mr-2 btn-primary float-right jh-create-entity create-quiz-exercise" [routerLink]="['export']">
+        <button *ngIf="routerContainPath()" class="btn mr-2 btn-primary float-right jh-create-entity create-quiz-exercise" [routerLink]="['new']">
+            <span class="fa fa-plus"></span>
```
Comment (rpadovani) [inline src/main/webapp/app/entities/quiz-exercise/quiz-exercise.component.html]: As above, please use `<fa-icon icon="plus">`

File: src/main/webapp/app/entities/quiz-exercise/quiz-exercise.component.html [hunk 14]
```diff
@@ -1,20 +1,34 @@
...
+            <span jhiTranslate="arTeMiSApp.quizExercise.home.createLabel">
+                Create new Quiz
+            </span>
+        </button>
+        <button *ngIf="!routerContainPath()" class="btn mr-2 btn-primary float-right jh-create-entity create-quiz-exercise" [routerLink]="['quiz-exercise', 'export']">
+            <span class="fa fa-plus"></span>
```
Comment (rpadovani) [inline src/main/webapp/app/entities/quiz-exercise/quiz-exercise.component.html]: As above :)

File: src/main/webapp/app/entities/quiz-exercise/quiz-exercise.component.ts [hunk 15]
```diff
@@ -51,10 +51,12 @@ export class QuizExerciseComponent implements OnInit, OnDestroy {
     }
 
     load() {
-        this.subscription = this.route.params.subscribe(params => {
-            this.courseId = params['courseId'];
-            this.loadForCourse(this.courseId);
-        });
+        if (this.course == null) {
+            this.subscription = this.route.params.subscribe(params => {
```
Comment (rpadovani) [inline src/main/webapp/app/entities/quiz-exercise/quiz-exercise.component.ts]: Do you need a subscription here? You never unsubscribe from it

File: src/main/webapp/app/entities/text-exercise/text-exercise.component.ts [hunk 16]
```diff
@@ -36,10 +37,10 @@ export class TextExerciseComponent implements OnInit, OnDestroy {
...
 
     ngOnInit() {
-        this.subscription = this.route.params.subscribe(params => {
-            this.load();
-            this.registerChangeInTextExercises();
-        });
+            this.subscription = this.route.params.subscribe(params => {
+                this.load();
+                this.registerChangeInTextExercises();
+            });
```
Comment (jpbernius) [inline src/main/webapp/app/entities/text-exercise/text-exercise.component.ts]: Please revert

File: src/main/webapp/app/entities/text-exercise/text-exercise.component.ts [hunk 17]
```diff
@@ -65,9 +66,7 @@ export class TextExerciseComponent implements OnInit, OnDestroy {
         });
     }
 
-    ngOnDestroy() {
-        this.eventManager.destroy(this.eventSubscriber);
-    }
+
```
Comment (jpbernius) [inline src/main/webapp/app/entities/text-exercise/text-exercise.component.ts]: Remove empty lines

File: src/main/webapp/app/entities/course/course.route.ts [hunk 18]
```diff
@@ -39,37 +39,37 @@ export const courseRoute: Routes = [
         canActivate: [UserRouteAccessService]
     },
     {
-        path: 'course/:id/view',
-        component: CourseDetailComponent,
+        path: 'coursenew',
```
Comment (jpbernius) [inline src/main/webapp/app/entities/course/course.route.ts]: What was wrong with `'course/new'`? I like the old one better.

File: src/main/webapp/app/entities/programming-exercise/programming-exercise.component.html [hunk 19]
```diff
@@ -8,7 +8,13 @@ <h4>
...
             </span>
         </button>
-        <button *ngIf="course" id="jh-create-entity" class="btn btn-primary float-right jh-create-entity create-programming-exercise mr-1" [routerLink]="['programming-exercise', 'new']">
+        <button *ngIf="course && !routerContainPath()" id="jh-create-entity" class="btn btn-primary float-right jh-create-entity create-programming-exercise mr-1" [routerLink]="['programming-exercise','new']">
+            <fa-icon [icon]="'plus'"></fa-icon>
+            <span class="hidden-sm-down" jhiTranslate="arTeMiSApp.programmingExercise.home.createLabel">
+            Create new Programming Exercise
+            </span>
+        </button>
+        <button *ngIf="course && routerContainPath()" id="jh-create-entity" class="btn btn-primary float-right jh-create-entity create-programming-exercise mr-1" [routerLink]="['new']">
```
Comment (jpbernius) [inline src/main/webapp/app/entities/programming-exercise/programming-exercise.component.html]: There needs to be a different way of doing this!

File: src/main/webapp/app/entities/exercise/exercise.component.ts [hunk 20]
```diff
@@ -7,13 +7,18 @@ import { Course, CourseService } from 'app/entities/course';
...
     @Input() embedded = false;
+    @Input() course: Course;
+    @Output() exerciseCount = new EventEmitter<number>();
     showAlertHeading: boolean;
     showHeading: boolean;
     courseId: number;
-    @Input() course: Course;
-    @Output() exerciseCount = new EventEmitter<number>();
+    predicate: string;
+    reverse: boolean;
```
Comment (jpbernius) [inline src/main/webapp/app/entities/exercise/exercise.component.ts]: Thanks for starting this refactoring! Can we move the methods also?
